src/AElf.Kernel.CodeCheck/Application/CodeCheckValidationProvider.cs

```diff
@@ -52,18 +54,37 @@ public async Task<bool> ValidateBlockAfterExecuteAsync(IBlock block)
             BlockHeight = block.Header.Height,
             ContractAddress = genesisContractAddress
         }).GetContractCodeHashListByDeployingBlockHeight.CallAsync(new Int64Value { Value = block.Header.Height });
-
-        if (codeHashList == null || !codeHashList.Value.Any())
+        
+        foreach (var codeHash in codeHashList.Value)
         {
-            Logger.LogInformation("CodeHashList is empty.");
-            return true;
+            if (_checkedCodeHashProvider.IsCodeHashExists(new BlockIndex
+                {
+                    BlockHash = blockHash,
+                    BlockHeight = block.Header.Height
+                }, codeHash))
+            {
+                continue;
+            }
+
+            var contractRegistration = await _contractReaderFactory.Create(new ContractReaderContext
+            {
+                BlockHash = blockHash,
+                BlockHeight = block.Header.Height,
+                ContractAddress = genesisContractAddress
+            }).GetSmartContractRegistrationByCodeHash.CallAsync(codeHash);
+
+            if (await _codeCheckService.PerformCodeCheckAsync(contractRegistration.Code.ToByteArray(),
+                    blockHash, block.Header.Height, contractRegistration.Category,
+                    contractRegistration.IsSystemContract, contractRegistration.IsUserContract))
+            {
+                continue;
```

Comment:
Should we call `_checkedCodeHashProvider.AddCodeHashAsync` after validation?

Reply:
**Related to this, but not resolvable this time round:**

1. Should we call `_checkedCodeHashProvider.AddCodeHashAsync` inside `CodeCheckService.PerformCodeCheckAsync`?
2. Does it make sense to have `_checkedCodeHashProvider.AddCodeHashAsync`  depend on the block height?
3. Does it make sense to configure `RequiredAcs` in Configuration contract? Or should we make `RequiredAcs` backward compatible?

Reply:
> Should we call `_checkedCodeHashProvider.AddCodeHashAsync` after validation?

I think it may not be good to add it here because the original data is stored according to the height, the real height is not known here, and the current height may make the data confusing. College here has a legacy: there's no place to clear the data.

Reply:
> **Related to this, but not resolvable this time round:**
> 
> 1. Should we call `_checkedCodeHashProvider.AddCodeHashAsync` inside `CodeCheckService.PerformCodeCheckAsync`?
> 2. Does it make sense to have `_checkedCodeHashProvider.AddCodeHashAsync`  depend on the block height?
> 3. Does it make sense to configure `RequiredAcs` in Configuration contract? Or should we make `RequiredAcs` backward compatible?

1、It is not necessary to store code everywhere you do `CodeCheckService.PerformCodeCheckAsync`, like tx pool pre-validation. So `_checkedCodeHashProvider.AddCodeHashAsync` inside `CodeCheckService.PerformCodeCheckAsync` maybe not a good idea.